packages/core-api/src/apis/implementations/auth/oauth2/OAuth2.ts

```diff
@@ -0,0 +1,181 @@
+  idToken: string;
+  scope: string;
+  expiresInSeconds: number;
+};
+
+const DEFAULT_PROVIDER = {
+  id: 'oauth2',
+  title: 'Your Identity Provider',
+  icon: OAuth2Icon,
+};
+
+const SCOPE_PREFIX = '';
+
+class OAuth2
+  implements OAuthApi, OpenIdConnectApi, ProfileInfoApi, SessionStateApi {
+  static create({
+    apiOrigin,
+    basePath,
+    environment = 'development',
+    provider = DEFAULT_PROVIDER,
+    oauthRequestApi,
+  }: CreateOptions) {
+    const connector = new DefaultAuthConnector({
+      apiOrigin,
+      basePath,
+      environment,
+      provider,
+      oauthRequestApi: oauthRequestApi,
+      sessionTransform(res: OAuth2Response): OAuth2Session {
+        return {
+          profile: res.profile,
+          idToken: res.idToken,
+          accessToken: res.accessToken,
+          scopes: OAuth2.normalizeScopes(res.scope),
+          expiresAt: new Date(Date.now() + res.expiresInSeconds * 1000),
+        };
+      },
+    });
+
+    const sessionManager = new RefreshingAuthSessionManager({
+      connector,
+      defaultScopes: new Set([
+        'openid',
+        `${SCOPE_PREFIX}userinfo.email`,
+        `${SCOPE_PREFIX}userinfo.profile`,
+      ]),
+      sessionScopes: session => session.scopes,
+      sessionShouldRefresh: session => {
+        const expiresInSec = (session.expiresAt.getTime() - Date.now()) / 1000;
+        return expiresInSec < 60 * 5;
+      },
+    });
+
+    return new OAuth2(sessionManager);
+  }
+
+  private readonly sessionStateTracker = new SessionStateTracker();
+
+  sessionState$(): Observable<SessionState> {
+    return this.sessionStateTracker.observable;
+  }
+
+  constructor(private readonly sessionManager: SessionManager<OAuth2Session>) {}
+
+  async getAccessToken(
+    scope?: string | string[],
+    options?: AccessTokenOptions,
+  ) {
+    const normalizedScopes = OAuth2.normalizeScopes(scope);
+    const session = await this.sessionManager.getSession({
+      ...options,
+      scopes: normalizedScopes,
+    });
+    this.sessionStateTracker.setIsSignedId(!!session);
+    if (session) {
+      return session.accessToken;
+    }
+    return '';
+  }
+
+  async getIdToken(options: IdTokenOptions = {}) {
+    const session = await this.sessionManager.getSession(options);
+    this.sessionStateTracker.setIsSignedId(!!session);
+    if (session) {
+      return session.idToken;
+    }
+    return '';
+  }
+
+  async logout() {
+    await this.sessionManager.removeSession();
+    this.sessionStateTracker.setIsSignedId(false);
+  }
+
+  async getProfile(options: ProfileInfoOptions = {}) {
+    const session = await this.sessionManager.getSession(options);
+    this.sessionStateTracker.setIsSignedId(!!session);
+    if (!session) {
+      return undefined;
+    }
+    return session.profile;
+  }
+
+  static normalizeScopes(scopes?: string | string[]): Set<string> {
+    if (!scopes) {
+      return new Set();
+    }
+
+    const scopeList = Array.isArray(scopes)
```

Comment:
Should just return the `scopeList` at this point, the logic below is Google-specific.